posts/2021/07/moving-scipy-to-meson.md

```diff
@@ -0,0 +1,489 @@
+verbose. Using a few helper snippets to generate file listings in IPython
+saves time:
+```python
+>>> for f in sorted(os.listdir('.')):
+...     if f.endswith('.f'):
+...         print(f"  '{f}',")
+... 
+  'dgamln.f',
+  'zabs.f',
+  ...
+```
+
+*Meson has an imperative DSL with immutable objects.* The syntax of the DSL is
+inspired by Python, and is easy to read. Objects - which can be dependencies,
+generated files, built extensions, etc. - being immutable makes things easy
+to debug, but in some cases it can restrict what you can do. For example,
+this code for building all `scipy.sparse.csgraph` extensions is quite
+elegant, but the `foreach` pattern will not work if the `.pyx` files are
+generated. This is because they'd then be objects created by `custom_target`,
+and there is no syntax to give them unique names within the `foreach` loop -
+this cost me some time to figure out:
+```python
+pyx_files = [
+  ['_flow', '_flow.pyx'],
+  ['_matching', '_matching.pyx'],
+  ['_min_spanning_tree', '_min_spanning_tree.pyx'],
+  ['_reordering', '_reordering.pyx'],
+  ['_shortest_path', '_shortest_path.pyx'],
+  ['_tools', '_tools.pyx'],
+  ['_traversal', '_traversal.pyx'],
+]
+
+foreach pyx_file : pyx_files
+  py3.extension_module(pyx_file[0],
+    pyx_file[1],
+    include_directories : inc_np,
+    dependencies : py3_dep,
+    install : true,
+    subdir : 'scipy/sparse/csgraph')
+endforeach
+```
+
+*Meson uses `pkg-config` for discovering dependencies.* This means certain
+things "just work". E.g., `blas = dependency('openblas')` followed by using
+`dependencies: blas` to build the `scipy.linalg._fblas` extension worked on
+the first try for me - definitely did not expect that.
+
+*There is one escape hatch to do things Meson does not natively support: `custom_target`.*
+As long you list its source files and outputs, you can invoke Python scripts
+through `custom_target`. This is how the SciPy build invokes `numpy.f2py`,
+Pythran and `cython --cplus` (Meson's Cython support is brand new, and only
+supports targeting C if you give a `.pyx` file to `py3.extension_module`
+right now).
+
+
+## The sharp bits
+
+Not everything was smooth sailing. Here are a few of the most important
+issues I ran into:
+
+1. Cython + generated sources. This was by far the most time-consuming topic.
+   Cython's `cimport` mechanism relies on the complete source tree. For
+   example, not having a `__init__.py` file two directories up from where the
+   `.pyx` file that you are building lives changes the C code that Cython
+   generates. The full source tree layout matters for Cython. Cython was
+   largely designed around in-tree builds, and Meson only allows you to
+   generate files out-of-tree. This necessitates hacks like writing scripts
+   or invoking `cp` to copy over many `__init__.py` and `.pxi|.pxd` files.
+   And then more hacks to ensure those files are respected as dependencies by
+   Ninja.
```

Comment:
Maybe something to bring up with cython before the 3.0 release, could they add `__file__` to the cimport path?

Reply:
`__file__` is added to the cimport path. The issue is that the `cimport` in the `.pyx` file is now in `build/scipy/xxx/` rather than in `scipy/xxx/` and there's no way for Cython to know that the `.pyx` was code-generated. And the codegen script has no good way of adding `scipy/xxx/` to the import path.

Reply:
The whole `pyximport` mechanism is one big hack as far as I can tell.